File: 233.数字-1-的个数.cpp
/*
 * @lc app=leetcode.cn id=233 lang=cpp
 *
 * [233] 数字 1 的个数
 */

// @lc code=start
class Solution {
    #define mset(x,v) memset(x,(v),sizeof(x))
    #define rep(i,a,n) for(int i=(a);i<=(n);i++)
    #define ll long long
    #define maxn 15
    ll a[maxn],pw[maxn]={1},dp[maxn];
    ll dfs(int pos,int cnt,bool limit){
        if(!pos) return cnt;
        if(~dp[pos]&&!limit) return dp[pos]+cnt*pw[pos];
        ll ret=0,bound=limit?a[pos]:9;
        rep(i,0,bound) ret+=dfs(pos-1,cnt+(i==1),i==bound&&limit);
        if(!limit) dp[pos]=ret;
        return ret;
    }
    ll solve(ll x){
        int len=0;
        while(x) a[++len]=x%10,x/=10;
        mset(dp,-1);
        return dfs(len,0,1);
    }
public:
    int countDigitOne(int n) {
        rep(i,1,maxn-1) pw[i]=pw[i-1]*10;
        return solve(n);
    }
};

// @lc code=end

File: 233.数字-1-的个数.py
#
# @lc app=leetcode.cn id=233 lang=python3
#
# [233] 数字 1 的个数
#

# @lc code=start
class Solution:
    def __init__(self):
        self.dp = None   
        self.n = None
        self.pw = []

    def countDigitOne(self, n: int) -> int:
        n = [int(i) for i in str(n)]
        self.dp = [-1] * len(n)
        self.pw = list(reversed([10**i for i in len(i)]))
        return self.dfs(len(n),0,True)
        
    def dfs(self,pos,count,limit):
        if not pos:
            return count
        if not dp[pos] and not limit:
            return dp[pos] + count * self.pw[pos]
        return  = 0
        for i in range( self.n[pos] if limit else 9):
            ret += self.dfs(pos-1,count + (1 if i==1 else 0),limit and i==bound)
        if not limit:
            dp[pos] = ret 
        return ret        

    def getDightLength(self,n):
        res = 0
        while n:
            res += 1
            n //= 10
        return res


# @lc code=end

if __name__ == "__main__":
    # print(Solution().countDigitOne(13))
    print(Solution().getDightLength(123))